network/common/src/main/java/org/apache/spark/network/protocol/MessageEncoder.java

```diff
@@ -72,9 +80,84 @@ public void encode(ChannelHandlerContext ctx, Message in, List<Object> out) {
     in.encode(header);
     assert header.writableBytes() == 0;
 
-    out.add(header);
     if (body != null && bodyLength > 0) {
-      out.add(body);
+      out.add(new MessageWithHeader(header, headerLength, body, bodyLength));
+    } else {
+      out.add(header);
     }
   }
+
+  /**
+   * A wrapper message that holds two separate pieces (a header and a body) to avoid
+   * copying the body's content.
+   */
+  private static class MessageWithHeader extends AbstractReferenceCounted implements FileRegion {
```

Comment:
Right, it is not documented because nio is currently "fast enough" and because epoll in the past was less stable.